backend/lib/kubernetes-client/cache/Reflector.js

```diff
@@ -0,0 +1,339 @@
+    const pager = ListPager.create(this.listWatcher)
+    const options = {
+      resourceVersion: this.relistResourceVersion
+    }
+
+    if (!this.paginatedResult && options.resourceVersion !== '' && options.resourceVersion !== '0') {
+      pager.pageSize = 0
+    }
+
+    this.store.synchronizing()
+
+    let list
+    try {
+      list = await pager.list(options)
+    } catch (err) {
+      if (isExpiredError(err)) {
+        this.isLastSyncResourceVersionExpired = true
+        // Retry immediately if the resource version used to list is expired.
+        // The pager already falls back to full list if paginated list calls fail due to an "Expired" error on
+        // continuation pages, but the pager might not be enabled, or the full list might fail because the
+        // resource version it is listing at is expired, so we need to fallback to resourceVersion="" in all
+        // to recover and ensure the reflector makes forward progress.
+        try {
+          list = await pager.list({
+            resourceVersion: this.relistResourceVersion
+          })
+        } catch (err) {
+          logger.error('%s: Failed to call full list %s: %s', this.name, this.expectedTypeName, err)
+          return
+        }
+      }
+      logger.error('%s: Failed to call paginated list %s: %s', this.name, this.expectedTypeName, err)
+      return
+    }
+
+    const {
+      resourceVersion,
+      paginated: paginatedResult
+    } = list.metadata
+
+    // We check if the list was paginated and if so set the paginatedResult based on that.
+    // However, we want to do that only for the initial list (which is the only case
+    // when we set ResourceVersion="0"). The reasoning behind it is that later, in some
+    // situations we may force listing directly from etcd (by setting ResourceVersion="")
+    // which will return paginated result, even if watch cache is enabled. However, in
+    // that case, we still want to prefer sending requests to watch cache if possible.
+    //
+    // Paginated result returned for request with ResourceVersion="0" mean that watch
+    // cache is disabled and there are a lot of objects of a given type. In such case,
+    // there is no need to prefer listing from watch cache.
+    if (options.resourceVersion === '0' && paginatedResult) {
+      this.paginatedResult = true
+    }
+
+    this.isLastSyncResourceVersionExpired = false
+    this.store.replace(list.items)
+    this.lastSyncResourceVersion = resourceVersion
+    try {
+      while (!this.stopRequested) {
+        const options = {
+          allowWatchBookmarks: true,
+          timeoutSeconds: randomize(this.minWatchTimeout.asSeconds()),
+          resourceVersion: this.lastSyncResourceVersion
+        }
+        try {
+          this.socket = this.listWatcher.watch(options)
+          await pEvent(this.socket, 'open', { timeout: this.connectTimeout.asMilliseconds() })
+        } catch (err) {
+          if (isExpiredError(err)) {
```

Comment:
handle resource version too new / large

Reply:
I am not able to reproduce this situation, even if I set the resourceVersion to a value much higher than any resourceVersion can be. I think the error has been removed and no event is emitted because the to high resourceVersion is much higher than the one of the changed object.

Reply:
Based on the go implementation https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/errors.go#L176-L182 we handle this error now after the watch socket has been opened but received  the corresponding StatusError. In this case we do a retry after the given time with the same resource version. The established socket is closed.